New Rule: object has acessible name (8f3b6) (#1279)
act-rules/act-rules.github.io: object-has-acessible-name-8f3b6.md

---

> id: 8fc3b6

Reviewer: The filename is incorrect (8f3b6 instead of 8fc3b6).

name: Object element in the accessibility tree has non-empty accessible name
rule_type: atomic
description: |

>   This rule checks that each object has an accessible name.

Reviewer: ```suggestion
  This rule checks that each object has a non-empty accessible name.
```

accessibility_requirements:
  wcag20:1.1.1: # Non-text Content (A)
    forConformance: true
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed
input_aspects:
  - DOM Tree
  - CSS Styling
  - Accessibility Tree

> acknowledgments:

Reviewer: Maybe add the source of the Moon speech used in the examples.

Author: This file has already been used in other rules so I don't know who to acknowledge.

Reviewer: Something like "audio is excerpt from JFK's 'Moon Speech'".

  authors:
    - António Estriga
---

## Applicability

The rule applies to any `object` element that is [included in the accessibility tree][].

## Expectation

Each target element has an [accessible name][] that is not empty (`""`).

**Note:** Testing that the [accessible name][] describes the purpose of the element is not part of this rule and must be tested separately.

## Assumptions

> There are currently no assumptions

Reviewer: ```suggestion
_There are currently no assumptions._
```

## Accessibility Support

Non supported media formats make screen readers render the text content of the element instead of other attributes.

## Background

- [Understanding Success Criterion 1.1.1: Non-text Content](https://www.w3.org/WAI/WCAG21/Understanding/non-text-content.html>)

## Test Cases

### Passed

#### Passed Example 1

> This `object` element has a non-empty [accessible name][] through aria-label.

Reviewer: ```suggestion
This `object` element has a non-empty [accessible name][] through its `aria-label` attribute.
```

```html
<object
  aria-label="Moon speech"
  data="/test-assets/moon-audio/moon-speech.mp3"
></object>
```

#### Passed Example 2

> This `object` element has a non-empty [accessible name][] through title.

Reviewer: ```suggestion
This `object` element has a non-empty [accessible name][] through its `title` attribute.
```

```html
<object
  title="Moon speech"
  data="/test-assets/moon-audio/moon-speech.mp3"
></object>
```

#### Passed Example 3

> This `object` element has a non-empty [accessible name][] through aria-labelledby.

Reviewer: ```suggestion
This `object` element has a non-empty [accessible name][] through its `aria-labelledby` attribute.
```

```html
<span id="label">Moon speech</span>
<object
  aria-labelledby="label"
  data="/test-assets/moon-audio/moon-speech.mp3"
></object>
```

#### Passed Example 4

This `object` element placed off screen has a non-empty [accessible name][].

```html
<html>
  <style>
    .offScreen {
      position: absolute;
      left: -9999px;
      top: -9999px;
    }
  </style>
  <body>
    <object
      title="Moon speech"
      data="/test-assets/moon-audio/moon-speech.mp3"
      class="offScreen"
    ></object>
  </body>
</html>
```

### Failed

#### Failed Example 1

This `object` element has an empty [accessible name][].

```html
<object title="" data="/test-assets/moon-audio/moon-speech.mp3"></object>
```

#### Failed Example 2

This `object` element has an empty [accessible name][].

```html
<span id="label"></span>
<object
  aria-labelledby="label"
  data="/test-assets/moon-audio/moon-speech.mp3"
></object>
```

#### Failed Example 3

> This `object` element has an empty [accessible name][] because the `aria-labelledby` references a non-existing id.

Reviewer: ```suggestion
This `object` element has an empty [accessible name][] because the `aria-labelledby` attribute references a non-existing id.
```

```html
<object
  aria-labelledby="download"
  data="/test-assets/moon-audio/moon-speech.mp3"
></object>
```

#### Failed Example 4

This `object` element has an empty [accessible name][].

```html
<object data="/test-assets/moon-audio/moon-speech.mp3"></object>
```

> ### Inapplicable

Reviewer: Can you add an inapplicable example where there is no `object` element?
